keps/sig-scheduling/3022-min-domains-in-pod-topology-spread/README.md

```diff
@@ -0,0 +1,468 @@
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md)
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+With [Pod Topology Spread](/keps/sig-scheduling/895-pod-topology-spread), users can define the rule to spread pods across your cluster among failure-domains.
+And, we propose to add the parameter `minDomains` to limit the minimum number of domains in Pod Topology Spread.
+
+## Motivation
+
+Pod Topology Spread has [`maxSkew` parameter](https://github.com/kubernetes/enhancements/tree/11a976c74e1358efccf251d4c7611d05ce27feb3/keps/sig-scheduling/895-pod-topology-spread#maxskew), which control the degree to which Pods may be unevenly distributed.
+But, there isn't a way to control the number of domains over which we should spread.
+
+### Goals
+
+Users can define a minimum number of domains with `minDomains` parameter.
+
+When the number of domains that have matching Pods is less than `minDomains`,
+
+- when `whenUnsatisfiable` equals `DoNotSchedule`, scheduler doesn't schedule a matching Pod to Nodes on the domains that have matching Pods.
+  i.e. scheduler filters Nodes on the domains that have matching Pods in the `Filter` phase.
+- when `whenUnsatisfiable` equals `ScheduleAnyway`, scheduler prefers Nodes on the domains that don't have matching Pods.
+  i.e. scheduler gives a high score to Nodes on the domains that don't have matching Pods in the `Score` phase.
+
+## Proposal
+
+### User Story
+
+I am using cluster autoscaler and I want to force spreading a deployment over at least 10 Nodes.
+
+## Design Details
+
+### API
+
+New parameter called `MinDomains` is introduced.
+
+```go
+type TopologySpreadConstraint struct {
+......
+  // MinDomains describes the minimum number of domains.
+  // When the number of domains that have matching Pods is less than `minDomains`,
+  // - when `whenUnsatisfiable` equals `DoNotSchedule`, scheduler doesn't schedule a matching Pod to Nodes on the domains that have matching Pods.
```

Comment:
Uhm... rethinking a bit more about this, I think the semantics are off.

Let's say maxSkew=2 and there is only one node. I should be able to schedule 2 pods in that one node. The third pod should fail for force a scale up.

So we have to consider how many domains have pods. If it's less than minDomains, the global minimum is 0 https://kubernetes.io/docs/concepts/workloads/pods/pod-topology-spread-constraints/#spread-constraints-for-pods

Reply:
Aldo is right. Currently (without `minDomains`), the semantics of "global minimum" literally indicates the minimum number of matching pods among all topology domains.

While with `minDomains`, the semantics is sort of dynamic:

- if the number of domains >= `minDomains`, "global minimum" stay the same as before
- if the number of domains < `minDomains`, "global minimum" is treated as 0

Reply:
Thanks for the clear explanation.

Okay, what I thought was 
- `maxSkew` does not affect the behavior of `minDomains`.
- `minDomains` tries to increase the number of domains as much as possible. So, if the number of domains < `minDomains`, it always tries to bind matching Pods to the domains that don't have any matching Pods.

But, I also think your opinion was better. I'll fix the description. 🙏